add Oxli datatypes

mr-c opened the pull request:
Adds the binary datatypes complete with sniffers for the khmer/oxli project.

See https://github.com/galaxyproject/tools-iuc/pull/80 for the tool wrappers that use these.

The file formats are defined at http://khmer.readthedocs.org/en/latest/dev/binary-file-formats.html and there is a unix `file` "magic" at https://github.com/dib-lab/khmer/blob/master/lib/magic

mr-c committed:
add Oxli datatypes

mr-c commented:
Do I need to fix the flake8 errors in files I didn't touch? Ex: `/lib/galaxy/webapps/tool_shed/controllers/admin.py`

nsoranzo commented:
@mr-c No, don't worry, someone will fix it in a separate PR.

mr-c commented:
Thanks @nsoranzo 

This PR is ready for review & merging.

bgruening commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -908,3 +908,83 @@ def sniff( self, filename ):
+        OxliBinary.__init__(self, **kwd)
+
+    def sniff(self, filename):
+        return OxliBinary.sniff(self, filename, "02")
```
I'm not sure, but maybe we should use super() here?

Depending on the filesize of an example dataset a doctest would be nice:

```
    """
    OxliNodeGraph starts with magic string "OXLI01"
    >>> from galaxy.datatypes.sniff import get_test_fname
    >>> fname = get_test_fname( 'sequence.csfasta' )
    >>> OxliNodeGraph().sniff( fname )
    False
    >>> fname = get_test_fname( 'khmer.oxli01' )
    >>> OxliNodeGraph().sniff( fname )
    True
    """
```

mr-c replied:
These classes are very specific to Galaxy binary file type sniffing so I'm not concerned about future-proofing for complicated Object hierarchies :-)

I've never worked with Doctests before, this was fun!

bgruening commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -908,3 +908,83 @@ def sniff( self, filename ):
+    def sniff(self, filename):
+        return OxliBinary.sniff(self, filename, "06")
+
+Binary.register_sniffable_binary_format("oxli.countgraph", "oxlicg", OxliCountGraph)
```
I would prefer to have the register step directly under each class to stick to other definitions.

mr-c replied:
Done

bgruening commented:
@mr-c great stuff! Looking forward to get khmer in :)

mr-c committed:
added doctests + 1 fix + cleanup

Made helper function a proper @staticmethod

fixed error in sniffer (thank you doctests)

Moved registration lines after their respective classes

Added test data

mr-c commented:
Thanks for the feedback @bgruening , the doctests caught an error in my code :-)

Should I move the oxli sniffers up higher on the list? The "OXLI" magic string is very unique and if some of the sniffers are time consuming it would be nice to have ours run sooner.

I renamed one of the classes but forgot to update `config/datatypes_config.xml.sample` and the tests still pass. Any pointers on how to write a test to catch that?

General, trying-to-be helpful questions:
A lot of the binary formats have `__init__` methods that do nothing more than the default. Would you like them removed in another PR?

Seems to be a bit of duplication between the code and the config:
Is specifying the mimetype in `config/datatypes_config.xml.sample` redundant since Binary defines its own mimetype?

To both encourage tools to publish their `file` "magic" detection and reward those that already do it. it would be great to be able to make use of  `file` "magic" detection and keep just galaxy specific config and methods.

mr-c committed:
OxliLabelSet->OxliGraphLabels even in the config

mr-c committed:
drop trailing whitespace

mr-c commented:
Okay, the build failure is still due to the other formatting problems.

Should I squash this PR down to one commit?

dannon commented:
No need to squash.  Build can be rerun w/ #722 and it should pass.

bgruening commented:
> Thanks for the feedback @bgruening , the doctests caught an error in my code :-)

Sure, no problem.

> Should I move the oxli sniffers up higher on the list? The "OXLI" magic string is very unique and if some of the sniffers are time consuming it would be nice to have ours run sooner.

I would leave them more to the end of the list. They are not (yet) so much used.

> I renamed one of the classes but forgot to update `config/datatypes_config.xml.sample` and the tests still pass. Any pointers on how to write a test to catch that?

No sorry.

> General, trying-to-be helpful questions:
> A lot of the binary formats have `__init__` methods that do nothing more than the default. Would you like them removed in another PR?

Separate PR seems like a good idea to me.

> Seems to be a bit of duplication between the code and the config:
> Is specifying the mimetype in `config/datatypes_config.xml.sample` redundant since Binary defines its own mimetype?

Not every datatype defines a mimetype. But this can probably improved. @blankenberg any idea?

> To both encourage tools to publish their `file` "magic" detection and reward those that already do it. it would be great to be able to make use of  `file` "magic" detection and keep just galaxy specific config and methods.

Can you elaborate here a little bit more? Do you have any EDAM ontology number for your file formats?
If you rebase your commit this error will go away. But I think it's fine as it is. I will wait for @blankenberg comments.

Thanks Michael!

mr-c commented on config/datatypes_conf.xml.sample:
```diff
@@ -412,6 +412,12 @@
     <datatype extension="xmfa" type="galaxy.datatypes.msa:MauveXmfa" display_in_upload="True" />
 
     <datatype extension="RData" type="galaxy.datatypes.binary:RData" display_in_upload="true" description="Stored data from an R session"/>
+    <datatype extension="oxlicg" type="galaxy.datatypes.binary:OxliCountGraph" mimetype="application/octet-stream" display_in_upload="true"/>
```
Is `mimetype="application/octet-stream"` here
necessary given that the supertype `Binary`'s `.get_mime()` returns `application/octet-stream`?

mr-c commented:
I've rebased onto `dev` and am ready for merge.

The magic discussion has been moved to another issue: https://github.com/galaxyproject/galaxy/issues/724
